### ISSUE TITLE
How to define the algorithm of L2_Pool2d? 

### ISSUE THREAD

**Issue by mingmingtasd** [webmachinelearning/webnn#278]
As you know, the algorithm of  L2_Pool2d is based on the 
Lp-normalization function which should be Y = **(X1^P + X2^P + ... + Xn^P) ^ (1/P)**.

But for L2_pool2d, I am not sure whether need to average the sum of elements as 
**Y =( (X1^2 + X2^2 + ... + Xn^2)/2) ^ (1/2)** or directly use Lp-normalization function as 
Y = **(X1^2 + X2^2 + ... + Xn^2) ^ (1/2)**.

I find two papers: https://sci-hub.yncjkj.com/10.1109/cvpr.2011.5995370 and https://sci-hub.yncjkj.com/10.1109/tcsvt.2015.2461978, they describe the Lp-normalization as below:

<img width="189" alt="Capture2" src="https://user-images.githubusercontent.com/52519066/163949663-af26d5ec-9aeb-4bb3-9fe2-a68bd517ca90.PNG">

<img width="205" alt="Capture" src="https://user-images.githubusercontent.com/52519066/163949688-3688d27b-29de-42e5-8398-08ba9e05d1e1.PNG">

So I confirm that **Lp-normalization function should be Y = (X1^P + X2^P + ... + Xn^P) ^ (1/P)**, but I am still not sure whether need to averge the sum of elements for L2_pool2d. I go through some framwork API spec and find the description as below:

**1. NNAPI [ANEURALNETWORKS_L2_POOL_2D](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a2fb636e30d8853f9fa1a395e30660e92):**
output[b, i, j, c] =
    sqrt(sum_{di, dj} pow(input[b, strides[1] * i + di, strides[2] * j + dj, c], 2) /
         sum(1))
**2. ONNX [LpPool](https://github.com/onnx/onnx/blob/main/docs/Operators.md#lppool):**
LpPool consumes an input tensor X and applies Lp pooling across the tensor according to kernel sizes, stride sizes, and pad lengths. Lp pooling consisting of computing the Lp norm on all values of a subset of the input tensor according to the kernel size and downsampling the data into the output tensor Y for further processing.
**3. OpenVINO: Not Supported**
**4. DML [DML_LP_POOLING_OPERATOR_DESC ](https://docs.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_lp_pooling_operator_desc):**
Computes the Lp-normalized value across the elements within the sliding window over the input tensor. The value of the P variable in the Lp-normalization function Y = (X1^P + X2^P + ... + Xn^P) ^ (1/P), where X1 to Xn representing each of the values within the sliding window. In common use cases, this value is either set to 1 or 2, representing either the L1 or L2 normalization respectively.

So it seems that NNAPI [ANEURALNETWORKS_L2_POOL_2D](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a2fb636e30d8853f9fa1a395e30660e92) should do averge. But after verifying on DML, DML [DML_LP_POOLING_OPERATOR_DESC ](https://docs.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_lp_pooling_operator_desc) doesn't averge. Thus the algroithm and implementation for l2_pool2d in these frameworks may be different.

**Comment by anssiko** [comment]
@huningxin, please report your proposed approach in this issue for the WG to review when you've discussed this issue with @mingmingtasd.

**Comment by fdwr** [comment]
> ...I am not sure whether need to average the sum of elements...

@mingmingtasd : What sources are you seeing that *average* elements before the `^ (1/p)`? Everything above appears consistent, including the papers, ONNX, DML, and NNAPI, all supporting Lebesgue Pooling for p = 2 as `Y = (X1^2 + X2^2 + ... + Xn^2) ^ (1/2)`. The `/ sum(1)` in NNAPI is odd but ignorable.

**Comment by mingmingtasd** [comment]
> > ...I am not sure whether need to average the sum of elements...
> 
> @mingmingtasd : What sources are you seeing that _average_ elements before the `^ (1/p)`? Everything above appears consistent, including the papers, ONNX, DML, and NNAPI, all supporting Lebesgue Pooling for p = 2 as `Y = (X1^2 + X2^2 + ... + Xn^2) ^ (1/2)`. The `/ sum(1)` in NNAPI is odd but ignorable.

Agree. The L2Pool2d should follow and based on the **Lp-normalization function**(Y = (X1^P + X2^P + ... + Xn^P) ^ (1/P)).

**Comment by fdwr** [comment]
@mingmingtasd Is there anything remaining unresolved on this one, or can it be closed?

**Comment by mingmingtasd** [comment]
> @mingmingtasd Is there anything remaining unresolved on this one, or can it be closed?

Let's close it, thanks.

**Comment by fujunwei** [comment]
TFLite also need to average the count of sum elements, [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/tflite/src/tensorflow/lite/kernels/internal/optimized/optimized_ops.h;l=3341?q=src%2Ftensorflow%2Flite%2Fkernels%2Finternal%2Foptimized%2Foptimized_ops.h) is the l2_pool2d's kernel, so How to keep compatibility between TFLite models and ONNX models?

**Comment by huningxin** [comment]
@fujunwei 

> TFLite also need to average the count of sum elements ((X1^2 + X2^2 + ... + Xn^2)/n) ^ (1/2)

As discussed before, l2 normalization should be calculated by `Y = (X1^2 + X2^2 + ... + Xn^2) ^ (1/2)`.  Would this be an issue of TFLite implementation? @reillyeon 

BTW, do we have implementation experience on CoreML's l2_pool? https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.pool.l2_pool @philloooo

**Comment by fdwr** [comment]
If TFLite defines it that way, that may be a useful operation, but it's something besides L2 pooling o_o.

```
function poolLebesgue(input, axes, windowDimensions, padding, dilations, exponent)
    return root(poolSum(pow(input, exponent), axes, windowDimensions, padding, dilations), exponent)
    // y = (x1^p + x2^p + ... + xn^p) ^ (1/p); y is reduced for each kernel
endfunction
```

**Comment by a-sully** [comment]
> 🤔 If TFLite defines it that way, that may be a useful operation, but it's something besides L2 pooling 👀.

Agreed. It seems like a TFLite WebNN backend will have to decompose `l2pool2d`... that begs the questions:

### 1. Is this decomposition expressible in WebNN?

Recall that we eventually need to clearly define all WebNN operators #462

> return root(poolSum(pow(input, exponent), axes, windowDimensions, padding, dilations), exponent)

- `root` (`sqrt`) and `pow` are existing WebNN operators 👍 
- `poolSum` is not. That being said, all of WebNN's pooling operators share a common pooling mechanism. **Let's specify the decomposition of this pooling mechanism as a first step towards #462?** (Also potentially related to #573)

### 2. Is this decomposition expressible in TFLite?

Anything is possible - especially if a device type is not mandated #749 - but potentially with severe performance cliffs, especially for non-CPU backends ([e.g.](https://www.tensorflow.org/lite/performance/gpu#supported_ops)), and especially if the operator implementation has to be hand-rolled.

FWIW #689 has the same issue: if at least two backends (i.e. Core ML and DML) have consistent behavior, then maybe that's okay (at least once discrepancies like #180 are resolved!)

**Comment by a-sully** [comment]
> especially if the operator implementation has to be hand-rolled

I just realized https://github.com/webmachinelearning/webnn/issues/180#issuecomment-2255216214 provides a decomposition of `l2pool2d` using `sqrt(conv2d(pow(input, 2), filterOfOnes, {dilations...}))` - thanks @fwdr!

**Comment by fdwr** [comment]
> `poolSum` is not [an existing WebNN operator]

@a-sully The more time I devote to thinking of expressing aggregate operators in terms of more fundamental operators, the more I realize some primitive ops (like `poolSum`) are missing. Even if not *directly* useful by themselves for ML, they *can* be useful for composition of others. At least for poolSum ([see here](https://fdwr.github.io/MachineLearningOperators/OperatorFormulas.html)), there's an easy decomposition using convolution:

```c
function poolSum(input, axes, windowDimensions, padding, strides, dilations)
    return poolGeneric(input, axes, windowDimensions, padding, strides, dilations, add, 0)
    // OR  convolve(input, filter = ones(windowDimensions), axes, windowDimensions, padding, strides, dilations)
endfunction
```

**Comment by huningxin** [comment]
@fdwr 





> ```c
> function poolSum(input, axes, windowDimensions, padding, strides, dilations)
>     return poolGeneric(input, axes, windowDimensions, padding, strides, dilations, add, 0)
>     // OR  convolve(input, filter = ones(windowDimensions), axes, windowDimensions, padding, strides, dilations)
> endfunction
> ```

Thanks for the decomposition! It's very helpful. IIUC, we may need to set the convolution groups to input channels, and make the all-ones filter in shape [groups, 1, windowDimensions.height, windowDimensions.width].

**Comment by a-sully** [comment]
> Well, these {ONNX, DML, NNAPI} agree (and I suspect CoreML too).

Not related to `l2Pool2d`, but some thoughts regarding operator compatibility across platforms:
- I understand that ONNX and DML _are different things_, but in these discussions, I consider them to be ~one backend. They're cooperating layers of the same stack :)
- [NNAPI is being deprecated in Android 15](https://developer.android.com/ndk/guides/neuralnetworks). It should not factor into decisions about WebNN

**Comment by fujunwei** [comment]
I filed [the issue](https://github.com/tensorflow/tensorflow/issues/73742) in TensorFlow, they also consider it's a issue of TFLite kernel implementation, maybe they will fix it later.

**Comment by fdwr** [comment]
> I filed [the issue](https://github.com/tensorflow/tensorflow/issues/73742) in TensorFlow, they also consider it's an issue of TFLite kernel implementation, maybe they will fix it later.

Thanks for filing. So it appears WebNN would need a decomposition until any future TFLite fix.

> potentially with severe performance cliffs ... especially if the operator implementation has to be hand-rolled

If it's any perf consolation, LP pooling is evidently not so common (only a few nodes in my little stash of hundreds of 1411 model files).

**Comment by huningxin** [comment]
@a-sully 

> * I understand that ONNX and DML _are different things_, but in these discussions, I consider them to be ~one backend. They're cooperating layers of the same stack :)

Regarding to our [new op proposal check list](https://github.com/webmachinelearning/webnn/blob/main/CONTRIBUTING.md#proposing-and-adding-a-new-operation), there are two aspects: Cross-framework support and Cross-platform implementability. I understand we usually study ONNX / ONNXRuntime as one example of frameworks, among TensorFlow and Pytorch etc., and investigate DML as one example of platform APIs, among TFLite and CoreML.

**Comment by a-sully** [comment]
> **Austin**: If it's any perf consolation, LP pooling is evidently not so common (only a few nodes in my little stash of hundreds of 1411 model files).

If that's the case and there's a straightforward decomposition which could be performed in "userspace"... is this operator needed in WebNN at all?

Put another way, if this operator wasn't already in the WebNN spec, would it pass the [new op proposal checklist](https://github.com/webmachinelearning/webnn/blob/main/CONTRIBUTING.md#proposing-and-adding-a-new-operation)?

**Comment by fdwr** [comment]
> It seems like a TFLite WebNN backend will have to decompose l2pool2d..
> ...
> copybara-service bot mentioned this issue 3 days ago

Woot, it appears that TFLite is already fixed (per https://github.com/tensorflow/tensorflow/pull/74079/files), which means it's just contingent on Chromium updating it's TF version. Given there will now be a direct call to TFLite with no decomposition, does that change the difficulty of implementing this?

> is this operator needed in WebNN at all? ...
> if this operator wasn't already in the WebNN spec, would it pass the [new op proposal checklist](https://github.com/webmachinelearning/webnn/blob/main/CONTRIBUTING.md#proposing-and-adding-a-new-operation)?

I think it's still would worth adding to the complete collective of pooling operations, and several implementations offer faster implementations than their decomposition, suggesting it's useful (even if it's rare in my mini-model collection, I do see people asking questions about it on forums). Additionally (unlike with localResponseNormalization where multiple implementations have little complicating variations), all the major libraries have an L2 pooling implementation and implement it consistently (barring that one implementation bug which is now fixed).

- *Use Case. What user scenarios or experiences will benefit from this operation, and why aren't existing operations sufficient?* Computer vision and pattern recognition.
- *Sample models. What are specific models that enable the target use case? One or more sample models as references are required.*  VGG with L2 pooling in Geodesics of learned representations https://arxiv.org/abs/1511.06394
- *Cross-framework support. Is an identical or similar operation supported by multiple popular frameworks? What are they?* Yes.
- *Cross-platform implementability. Is the operation implementable in more than one platform? What are they?* Yes.

**Comment by a-sully** [comment]
Thanks for thoroughly following through with this issue @fdwr. TFLite's alignment with the other platforms does improve the "Cross-framework support" line item. Seems reasonable to me 👍 Filed https://crbug.com/361717758 to track implementation in Chromium

Can we close this issue?

**Comment by fdwr** [comment]
Mingming, I'm closing it from the spec perspective, as Austin created a Chromium issue for it. 👍 Thank you for raising it.

**Comment by reillyeon** [comment]
No, it means that the internal change stalled and the automation closed the PR. I'm poking the author.

**Comment by fdwr** [comment]
> it means that the internal change stalled and the automation closed the PR. I'm poking the author.

Seems the latest progress for TFLite/LiteRT's L2Pool operator is here (https://github.com/tensorflow/tensorflow/pull/96599), with some tests failing because of a team expecting the current behavior (inconsistent with other libraries). So TFLite might need another operator for compat to unblock L2Pool to have the normal behavior.
